modules/gateway/peersmanager.go

```diff
@@ -73,6 +74,29 @@ func (g *Gateway) permanentPeerManager(closedChan chan struct{}) {
 	connectionLimiterChan := make(chan struct{}, maxConcurrentOutboundPeerRequests)
 
 	g.log.Debugln("INFO: [PPM] Permanent peer manager has started")
+
+	// Collect a list of peers that we have previously connected to.
+	g.mu.RLock()
+	var previousOutboundPeers []modules.NetAddress
+	for _, node := range g.nodes {
+		if node.WasOutboundPeer {
+			previousOutboundPeers = append(previousOutboundPeers, node.NetAddress)
+		}
+	}
+	g.mu.RUnlock()
+
+	// Construct a generator to return the next node considered for
+	// connection. The generator will first draw randomly from
+	// previousOutboundPeers, and then fallback to random node selection.
+	perm := fastrand.Perm(len(previousOutboundPeers))
+	nextNode := func() (addr modules.NetAddress, err error) {
+		if len(perm) > 0 {
+			addr, perm = previousOutboundPeers[perm[0]], perm[1:]
+			return addr, nil
+		}
+		return g.randomNode()
+	}
```

Comment:
both sources? Isn't `nextNode()` only called in one place?

More importantly though, you actually want the loop that creates `previousOutboundPeers` to be inside the inifinite loop of this function (line 100). That's because it runs for a long time, potentially days, during which your outbound peers may go online and offline frequently.

Currently, the `previousOutboundPeers` loop will only ever collect peers that you had as outbound when you first booted. But it's very likely that you'll pick up a few more outbound peers as you run, and you want to make sure they are added to the current pool you are selecting from.